### PR TITLE
Turn off unnecessary/ bad rules

### PR DESCRIPTION
**Pull request by lemonmade**
As we are slowly [turning on all ESLint rules in web](https://github.com/Shopify/web/issues/2073), a few have stuck out as being bad/ unnecessary. This PR turns them off, and I have commented on why I feel they should be turned off inline.

### CODE REVIEW

File: lib/config/node.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
 module.exports = {
   env: {
     node: true,
   },
 
   plugins: ['node'],
 
-  rules: require('./rules/node'),
+  rules: merge(require('./rules/node'), {
+    'no-process-env': 'off',
```
Comment (lemonmade) [inline lib/config/node.js:11]: Servers often need to respond to the environment.

File: lib/config/react.js [hunk 2]
```diff
@@ -21,7 +21,6 @@ module.exports = {
 
   rules: merge(require('./rules/react'), require('./rules/jsx-a11y'), {
     'shopify/react-initialize-state': 'error',
-    'shopify/react-type-state': 'error',
```
Comment (lemonmade) [inline lib/config/react.js:24]: We no longer need this rule, as TS has updated to make it so the that `state` is correctly checked against its type argument. I guess I should just delete the rule altogether?

File: lib/config/rules/promise.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ module.exports = {
   // Avoid wrapping values in Promise.resolve or Promise.reject when not needed
   'promise/no-return-wrap': 'error',
   // Ensure that inside a then() you make sure to return a new promise or value.
-  'promise/always-return': 'error',
+  'promise/always-return': 'off',
```
Comment (lemonmade) [inline lib/config/rules/promise.js:12]: IMO it is not a bad practice to simply do a side effect-ey thing after a promise resolves. This rule also creates inconsistencies in what is allowed for promises versus what is allowed for async await (we do not require those functions to return a value, even though they are just promises under the hood)

File: lib/config/rules/react.js [hunk 4]
```diff
@@ -34,9 +34,9 @@ module.exports = {
   // Prevent usage of deprecated methods
   'react/no-deprecated': 'error',
   // Prevent usage of setState in componentDidMount
-  'react/no-did-mount-set-state': 'error',
+  'react/no-did-mount-set-state': 'off',
   // Prevent usage of setState in componentDidUpdate
-  'react/no-did-update-set-state': 'error',
+  'react/no-did-update-set-state': 'off',
```
Comment (lemonmade) [inline lib/config/rules/react.js]: In both cases there is often no alternative for updating in response to prop changes/ doing something only on the client. This will be less of an issue with the new props from state stuff, but I don't feel these are particularly bad practices.

File: lib/config/rules/variables.js [hunk 5]
```diff
@@ -20,7 +20,7 @@ module.exports = {
   // Disallow use of undeclared variables unless mentioned in a /*global */ block
   'no-undef': 'error',
   // Disallow use of undefined variable
-  'no-undefined': 'error',
+  'no-undefined': 'off',
```
Comment (lemonmade) [inline lib/config/rules/variables.js:23]: I think this is a kind of antiquated rule. It is also very hard to follow in TS when we often type things as being optional for which `undefined` is the only allowable value to "unset" the field.

File: lib/config/webpack.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 const merge = require('merge');
 
 module.exports = {
-  rules: merge(require('./rules/webpack')),
+  rules: merge(require('./rules/webpack'), {
+    'no-process-env': 'off',
```
Comment (lemonmade) [inline lib/config/webpack.js:5]: Using this in webpack helps actually kill code that only applies in a particular environment.

File: lib/config/react.js (hunk 2 above)
Comment (ismail-syed) [reply lib/config/react.js:24]: 👍 for 🔥 the rule - no need to confuse our future selves of this not being used.

**Review by ismail-syed — APPROVED** [review]
👍 LGTM. 

Can we document these changes in the `CHANGELOG.md`, please. Not why we're changing them but the fact that we've disabled them.

**Review by kaelig — APPROVED** [review]
Looks good to me (please update the changelog before merging).
